In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('household_power_consumption.csv')

In [3]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,16/12/2006,17:24:00,4.216,0.418,234.84,18.4,0,1,17.0
1,16/12/2006,17:25:00,5.36,0.436,233.63,23,0,1,16.0
2,16/12/2006,17:26:00,5.374,0.498,233.29,23,0,2,17.0
3,16/12/2006,17:27:00,5.388,0.502,233.74,23,0,1,17.0
4,16/12/2006,17:28:00,3.666,0.528,235.68,15.8,0,1,17.0


#### the Reading is per minute and b/w (dec-2006 to dec-2008)

##### last rows

In [4]:
df.tail()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
1048570,13/12/2008,21:34:00,0.426,0.076,242.27,1.8,0,0,0.0
1048571,13/12/2008,21:35:00,0.424,0.076,242.1,1.8,0,0,0.0
1048572,13/12/2008,21:36:00,0.422,0.076,241.73,1.8,0,0,0.0
1048573,13/12/2008,21:37:00,0.422,0.078,242.56,1.8,0,0,0.0
1048574,13/12/2008,21:38:00,0.422,0.078,242.61,1.8,0,0,0.0


In [5]:
df.shape

(1048575, 9)

#### there are  data( 1048575 rows , 9 columns)

In [6]:
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 1048575 entries, 0 to 1048574
Data columns (total 9 columns):
 #   Column                 Non-Null Count    Dtype  
---  ------                 --------------    -----  
 0   Date                   1048575 non-null  str    
 1   Time                   1048575 non-null  str    
 2   Global_active_power    1048575 non-null  str    
 3   Global_reactive_power  1048575 non-null  str    
 4   Voltage                1048575 non-null  str    
 5   Global_intensity       1048575 non-null  str    
 6   Sub_metering_1         1048575 non-null  str    
 7   Sub_metering_2         1048575 non-null  str    
 8   Sub_metering_3         1044506 non-null  float64
dtypes: float64(1), str(8)
memory usage: 107.7 MB


#### all data are in "string" but only "Sub_metering_3" is float

#### data is lying in range (1 ,1048575)

In [7]:
df.describe()

,Sub_metering_3
count,1.044506e+06
mean,5.933837e+00
std,8.210398e+00
min,0.000000e+00
25%,0.000000e+00
50%,0.000000e+00
75%,1.700000e+01
max,3.100000e+01


In [8]:
df.describe(include = 'str' )

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2
count,1048575,1048575,1048575,1048575,1048575,1048575,1048575,1048575
unique,729,1440,4076,492,2605,219,82,80
top,17/12/2006,17:24:00,0.218,0,?,1,0,0
freq,1440,729,6714,266300,4069,95350,958708,746737


In [9]:
df.dtypes

Date                         str
Time                         str
Global_active_power          str
Global_reactive_power        str
Voltage                      str
Global_intensity             str
Sub_metering_1               str
Sub_metering_2               str
Sub_metering_3           float64
dtype: object

#### data type conversion`

In [10]:
df.isnull().sum()

Date                        0
Time                        0
Global_active_power         0
Global_reactive_power       0
Voltage                     0
Global_intensity            0
Sub_metering_1              0
Sub_metering_2              0
Sub_metering_3           4069
dtype: int64

### change datatype of Global_active_power , Global_reactive_power, Voltage, Global_intensity ,  in float

#### checkinig the row having "?"

In [11]:
df[df['Global_active_power'] == "?"]

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
6839,21/12/2006,11:23:00,?,?,?,?,?,?,NaN
6840,21/12/2006,11:24:00,?,?,?,?,?,?,NaN
19724,30/12/2006,10:08:00,?,?,?,?,?,?,NaN
19725,30/12/2006,10:09:00,?,?,?,?,?,?,NaN
41832,14/1/2007,18:36:00,?,?,?,?,?,?,NaN
...,...,...,...,...,...,...,...,...,...
1043669,10/12/2008,11:53:00,?,?,?,?,?,?,NaN
1043670,10/12/2008,11:54:00,?,?,?,?,?,?,NaN
1043671,10/12/2008,11:55:00,?,?,?,?,?,?,NaN
1043672,10/12/2008,11:56:00,?,?,?,?,?,?,NaN


#### now drop row having "?"

In [12]:
df[df['Global_active_power'] == "?"].shape

(4069, 9)

#### 1. As, we can see there are 4069 row which are having complete"?"
2. so . we droppiing

In [13]:
df = df[df['Global_active_power'] != "?"]

checking data type

In [14]:
df[df['Global_active_power'] == "?"]

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3


changing all other in float

In [15]:
df.columns

Index(['Date', 'Time', 'Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3'],
      dtype='str')

In [16]:
num_data = ['Global_active_power', 'Global_reactive_power',
       'Voltage', 'Global_intensity', 'Sub_metering_1', 'Sub_metering_2',
       'Sub_metering_3']

In [17]:
for i in num_data:
    df[i] = df[i].astype("float")

In [18]:
df.dtypes

Date                         str
Time                         str
Global_active_power      float64
Global_reactive_power    float64
Voltage                  float64
Global_intensity         float64
Sub_metering_1           float64
Sub_metering_2           float64
Sub_metering_3           float64
dtype: object

#### coverting the date column into datetime

In [19]:
df["Date"] = pd.to_datetime(df["Date"])

C:\Users\atinp\AppData\Local\Temp\ipykernel_11936\936118274.py:1: UserWarning: Parsing dates in %d/%m/%Y format when dayfirst=False (the default) was specified. Pass `dayfirst=True` or specify a format to silence this warning.
  df["Date"] = pd.to_datetime(df["Date"])


In [20]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0


### extracting  day_name, month_name , year  from data


In [21]:
df["day"] = df["Date"].dt.day

In [22]:
df["day_name"] = df["Date"].dt.day_name()

In [23]:
df["year"] = df["Date"].dt.year

In [24]:
df["month_name"] = df["Date"].dt.month_name()

In [25]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,day,day_name,year,month_name
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,16,Saturday,2006,December
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,16,Saturday,2006,December
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,16,Saturday,2006,December
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,16,Saturday,2006,December
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,16,Saturday,2006,December


In [26]:
df.drop(columns = ["Date"])

,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,day,day_name,year,month_name
0,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,16,Saturday,2006,December
1,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,16,Saturday,2006,December
2,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,16,Saturday,2006,December
3,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,16,Saturday,2006,December
4,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,16,Saturday,2006,December
...,...,...,...,...,...,...,...,...,...,...,...,...
1048570,21:34:00,0.426,0.076,242.27,1.8,0.0,0.0,0.0,13,Saturday,2008,December
1048571,21:35:00,0.424,0.076,242.10,1.8,0.0,0.0,0.0,13,Saturday,2008,December
1048572,21:36:00,0.422,0.076,241.73,1.8,0.0,0.0,0.0,13,Saturday,2008,December
1048573,21:37:00,0.422,0.078,242.56,1.8,0.0,0.0,0.0,13,Saturday,2008,December


### spliting the  hrs and mins from time columns

In [27]:
df[[ 'Hours' , 'Minutes','second']] = df["Time"].str.split(':' ,expand = True)

In [28]:
df.drop(columns = ["second"] , inplace =True)

In [29]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,day,day_name,year,month_name,Hours,Minutes
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,16,Saturday,2006,December,17,24
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,16,Saturday,2006,December,17,25
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,16,Saturday,2006,December,17,26
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,16,Saturday,2006,December,17,27
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,16,Saturday,2006,December,17,28


#### checking for the dupllicate row

In [30]:
df.duplicated().sum()

np.int64(0)

### check for the null values

In [31]:
df.isnull().sum()

Date                     0
Time                     0
Global_active_power      0
Global_reactive_power    0
Voltage                  0
Global_intensity         0
Sub_metering_1           0
Sub_metering_2           0
Sub_metering_3           0
day                      0
day_name                 0
year                     0
month_name               0
Hours                    0
Minutes                  0
dtype: int64

### create new col total metering

In [32]:
df['Total_metering'] = df[['Sub_metering_1','Sub_metering_2','Sub_metering_3']].sum(axis = 1)

In [33]:
df.head()

,Date,Time,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,day,day_name,year,month_name,Hours,Minutes,Total_metering
0,2006-12-16,17:24:00,4.216,0.418,234.84,18.4,0.0,1.0,17.0,16,Saturday,2006,December,17,24,18.0
1,2006-12-16,17:25:00,5.360,0.436,233.63,23.0,0.0,1.0,16.0,16,Saturday,2006,December,17,25,17.0
2,2006-12-16,17:26:00,5.374,0.498,233.29,23.0,0.0,2.0,17.0,16,Saturday,2006,December,17,26,19.0
3,2006-12-16,17:27:00,5.388,0.502,233.74,23.0,0.0,1.0,17.0,16,Saturday,2006,December,17,27,18.0
4,2006-12-16,17:28:00,3.666,0.528,235.68,15.8,0.0,1.0,17.0,16,Saturday,2006,December,17,28,18.0


In [34]:
df.drop(columns=['Date','Time'],inplace=True)

In [35]:
df.head()

,Global_active_power,Global_reactive_power,Voltage,Global_intensity,Sub_metering_1,Sub_metering_2,Sub_metering_3,day,day_name,year,month_name,Hours,Minutes,Total_metering
0,4.216,0.418,234.84,18.4,0.0,1.0,17.0,16,Saturday,2006,December,17,24,18.0
1,5.360,0.436,233.63,23.0,0.0,1.0,16.0,16,Saturday,2006,December,17,25,17.0
2,5.374,0.498,233.29,23.0,0.0,2.0,17.0,16,Saturday,2006,December,17,26,19.0
3,5.388,0.502,233.74,23.0,0.0,1.0,17.0,16,Saturday,2006,December,17,27,18.0
4,3.666,0.528,235.68,15.8,0.0,1.0,17.0,16,Saturday,2006,December,17,28,18.0


In [36]:
df.to_csv('household_cleandata.csv',index = False)

### Analysis